text/0000-mutex_traits.md

```diff
@@ -0,0 +1,174 @@
+- Feature Name: mutex_Traits
+- Start Date: 2015-06-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Introduce a distinction between `!Trait` and `?Trait` to enable mutually exclusive traits within the constraint that implementing a trait for a type should be a backwards compatible change. With these two type - trait relations distinguished from one another, allow negative impls of all traits as well as negative bounds, introducing mutual exclusion to the trait system. This enables users to encode additional logic in types & to provide multiple coherent blanket impls for parameterized types with mutually exclusive trait bounds, avoiding the backward compatibility problems that have hampered similar proposals in the past.
+
+# Motivation
+
+Trait coherence rules ensure that the compiler will predictably associate a particular impl with any invocation of a method or item associated with that trait. Without coherence rules, the behavior of a Rust program could easily become disastrously non-deterministic.
+
+However, the current coherence rules are very conservative in what they allow to be implemented. Other RFCs may introduce allowances for multiple overlapping implementations using an order of precendence for specialized implementations. This RFC addresses the limitation from another angle. By introducing mutually exclusive traits, it becomes possible to declare that two traits must not both be implemented by a single type, causing parameters bound by each of those traits to be non-overlapping.
+
+Take this system, which defines `Consumable`, `Edible` and `Poisonous` traits.
+
+```rust
+pub trait Consumable {
+    fn consume(&self);
+}
+
+pub trait Edible: !Poisonous {
+    fn nourish(&self);
+}
+
+pub trait Poisonous: !Edible {
```

Comment:
Do we require that both traits declare themselves to be negatively bounded by the other? It certainly improves readability, especially in larger source files. If this is the case, the compiler should generate a warning if `Edible : !Poisonous`, but `Poisonous` is not bounded by `!Edible`.